### ISSUE TITLE
setting GPU through "torch.set_default_tensor_type" and other details

### ISSUE THREAD

**Issue by adrienchaton** [KinWaiCheuk/nnAudio#8]
Hi ! I have installed your library and start running some trainings.
I have a couple of remarks from the first use.

You point torch.set_default_tensor_type as the way to run on GPU, it does work but it can cause some issues. For instance, in my case I cannot use set_default_tensor_type and multiple workers (it took me a bit to find that out): https://discuss.pytorch.org/t/cuda-initialization-error-when-dataloader-with-cuda-tensor/43390/3
I guess it would be better to have a way to set the device of the spectral operator itself. Such as an init argument or a function .to(device) ?

MelSpectrogram doesn't seem to have the arguments ,trainable_mel=__,trainable_STFT=__) that you mention in the readme. And sometimes it is a bit confused to guess if freq_bins,fmin,fmax can be let None for default setting or not.

Of course, it's at early development so it's great to have that in hands already. I just point out in case that helps. If you are interested, I can update you on if that does optimize as well as what I do so far (using torch.stft and librosa filters put on GPU), and if the other features do improve the fit of my models (I particularly look at the log scale STFT).

Best

**Comment by KinWaiCheuk** [comment]
> I guess it would be better to have a way to set the device of the spectral operator itself. Such as an init argument or a function .to(device) ?

Thanks for this useful suggestion. I have added the `device='cuda:0'` to each of the spectrogram method. 

> MelSpectrogram doesn't seem to have the arguments ,trainable_mel=,trainable_STFT=) that you mention in the readme. And sometimes it is a bit confused to guess if freq_bins,fmin,fmax can be let None for default setting or not.

I forgot to package the updated code to Pypi last time. Please install nnAudio 0.0.9 again, you should be able to use `trainable_mel` and `trainable_STFT` now. For the explanation of the arguments, I am trying to write a documentation, I am still learning how to use Sphinx to write my documentation.

>  I can update you on if that does optimize as well as what I do so far (using torch.stft and librosa filters put on GPU), and if the other features do improve the fit of my models (I particularly look at the log scale STFT).

It would be indeed a great help. I have been wanting to do a comparison on nnAudio and torch.stft. I had tried a quick test on [colab](https://colab.research.google.com/drive/1MMwpa9nvFlUly3aCnn1to9JeBKfXp94U), but it somehow shows that nnAudio is faster than torch.stft, which is out of my expectation since I expect that torch.stft is faster.

Anyhow, please install nnAudio version 0.0.9 and try again, I have updated a bit and fixed the above mentioned issues.

**Comment by adrienchaton** [comment]
Looking forward to the documentation. If you want a beta reader, I can do !
And I will definitely cite you if publishing with the model I currently work on.
Slight speed difference is not my point, it is more about training raw waveform generative models.
But the loss quality is very important !

One point, can you clarify the " added device='cuda:0' " please ?
How shall I myself adjust which GPU I will assign the nnAudio spectral operator to ? (our servers run with up to 8 GPUs so device='cuda:0' ... 'cuda:7')

**Comment by KinWaiCheuk** [comment]
For the `device` argument, it works like this

```python
spec_layer = Spectrogram.STFT(device='cuda:0') # create this STFT layer in cuda:0
spec_layer(x) # feed-forward
```

So when you create that spectrogram layer, you can now choose what GPU to use.

**Comment by adrienchaton** [comment]
Hi @KinWaiCheuk 
Thanks, I updated, the device argument does the good job.
However, it seems MelSpectrogram has not been updated, it does not accept the device argument and still do not let setting the trainable_mel/STFT.
Did you put it in the 0.0.9  update as well ?

**Comment by KinWaiCheuk** [comment]
I have done a quick test on Google Collab, it seems the `device` argument works fine
https://colab.research.google.com/drive/1UtswBYWhVxDNBRDajWzyplZfMiqENCEF

Just to points out that please use `Spectrogram.MelSpectrogram` instead of `Spectrogram.MelSpectrogramv2`. `Spectrogram.MelSpectrogramv2` is for debugging purpose, and I should remove it to avoid confusion next time when I push a new version.

Sorry for the confusions, I hope once I finish the documentation, everything will be more clear.

**Comment by adrienchaton** [comment]
alright, thanks it works ! I was importing MelSpectrogram and not using Spectrogram.MelSpectrogram

as it was shown in readme "Functionalities", maybe it is more clear to put both spectrogram and CQT examples as:
Spectrogram.MelSpectrogram()
Spectrogram.CQT()
